### PR TITLE
Drop support for EOL Python 2.7

### PR DESCRIPTION
**Pull request by hugovk**
This PR has the objective of dropping support for EOL Python 2.7 to reduce maintenance burden, reduce use of CI resources, modernise the codebase, and allow new Python features to be freely used.

Fixes https://github.com/petl-developers/petl/issues/478.

## Changes

1. Stop testing 2.7
2. Update docs, `python_requires` and classifiers
3. Remove redundant code such as `compat.py`
4. Update Python syntax using https://github.com/asottile/pyupgrade `----py36-plus`

## Checklist

Checklist for for pull requests including new code and/or changes to existing code...

* [ ] Code
  * [x] Includes unit tests
  * [ ] New functions have docstrings with examples that can be run with doctest
  * [ ] New functions are included in API docs
  * [ ] Docstrings include notes for any changes to API or behaviour
  * [x] All changes documented in docs/changes.rst
* [ ] Testing
  * [ ] \(Optional) Tested local against remote servers
  * [ ] Travis CI passes (unit tests run under Linux)
  * [x] AppVeyor CI passes (unit tests run under Windows)
  * [x] GitHub Actions CI passes (unit tests run under Linux, macOS, Windows)
  * [ ] Unit test coverage has not decreased (see Coveralls)
* [ ] Changes
  * [ ] \(Optional) Just a proof of concept
  * [ ] \(Optional) Work in progress
  * [x] Ready to review
  * [ ] Ready to merge

### CODE REVIEW

File: petl/comparison.py [hunk 1]
```diff
@@ -96,9 +90,9 @@ def __getitem__(self, item):
 
 def _typestr(x):
     # attempt to preserve Python 2 name orderings
-    if isinstance(x, binary_type):
+    if isinstance(x, bytes):
```
Comment (juarezr) [inline petl/comparison.py:93]: Wondering if the comment above is a py2 hack...

**Review by juarezr — COMMENTED** [review]
Great!
I'm wondering about what are your plans. 

Should this land in a major version change? like `2.0`?

Also, I would like to hear @alimanfoo about the recommended change.

File: petl/comparison.py (hunk 1 above)
Comment (hugovk) [reply petl/comparison.py:93]: Shall we refactor out `_typestr(x)` to `type(x).__name__`?

**Comment by hugovk** [comment]
> Should this land in a major version change? like `2.0`?

Some projects don't but generally they do bump the major version and I'd tend to agree, as it's a good way to signal for users.

**Comment by juarezr** [comment]
Some considerations:

1. As stated on @alimanfoo roadmap [here](https://groups.google.com/g/python-etl/c/0EOmiTOdrUE/m/c1vTuMGjDAAJ) it would be interesting to drop old behavior like #365 and [this](https://petl.readthedocs.io/en/stable/io.html#deprecated-i-o-helper-classes).
2. Other interesting migrations for v2.0 would be #559 and #453
3. Also, I wonder what could break within external libraries for:
   a. [File formats](https://petl.readthedocs.io/en/stable/io.html#xml-files)
   b. [Remote Access](https://petl.readthedocs.io/en/stable/io.html#remote-i-o-helper-classes)

File: petl/comparison.py (hunk 1 above)
Comment (juarezr) [reply petl/comparison.py:93]: - If the code is private and is not performance sensitive I wouldn't bother. 
- I didn't have the time for searching for similar cases through the changes

**Review by MihanixA — APPROVED** [review]

**Review by juarezr — CHANGES_REQUESTED** [review]
I think this is a great addition and the needed next step for `petl`.

But I wonder if also it would be possible before merging this PR:

- Make a compatibility break release, with a major version change to `2.0`.
- Change all the planned breakage previously issued/reported for:
  * Troublesome behavior change like #365 
  * Awkward API changes
  * Any modification that can impact end-users
- Executing extensive tests with [File formats](https://petl.readthedocs.io/en/stable/io.html#xml-files), [Databases](https://petl.readthedocs.io/en/stable/io.html#databases), and [Remote file access](https://petl.readthedocs.io/en/stable/io.html#remote-i-o-helper-classes)
- Evolve the possibility of maintaining a branch for some time to:
  * Tackling security vulnerabilities / CVE
  * Giving some time for users that can't afford to upgrade right now
- Hear the improvements that people are aiming for after the merge.
  * It's worth dropping python2.7 with no other planned improvement ?
  * Check if anyone would like to commit some effort for #453
  * Check if anyone would like to commit some time for #559 or another feature that depends on dropping `py2.x`.
- Hear project founder @alimanfoo opinion.

**Comment by hugovk** [comment]
Seems this has gotten out of sync, closing and it can be revisited afresh when the above issues are resolved.

**Comment by juarezr** [comment]
Yes, Sure.

The evolution in this PR paves the road for migration.
I think there are a few additional issues to solve and we are done.

**Pull request closed without merging**